App insights Namespace Filter

seanimam opened the pull request:
## Description
Adds namespace filters and namespace exceptions to App Insights Fluid Telemetry Logger. This updates the logic of the filter to be as such:
- If a `TelemetryFilter` specifies both a `category` and a `namespace`, the event must match both.
- If only a `category` or `namespace` is provided, the event should match either one of them.
- If a `namespace` pattern exception is specified in the `TelemetryFilter`, the event should not match the exception pattern.

Additionally, Tests were revamped to include more consistency, distinctly cover more edge cases and be easier to follow.

## Reviewer Guidance

seanimam committed:
Adds namespace filters and namespace exceptions

seanimam committed:
Adds jsdoc to doesEventMatchFilter method

Josmithr commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -39,15 +39,29 @@ export interface FluidAppInsightsLoggerConfig {
+	category?: TelemetryEventCategory;
+	/**
+	 * The namespace pattern to filter telemetry events.
+	 * For example, "perf:latency:*" will match any namespace starting with "perf:latency:"
```
There are a few things we can do here to leverage TSDoc a bit more. Structurally, I would recommend something like:

```typescript
/**
 * The namespace pattern to filter telemetry events.
 * @remarks Expressed as a RegEx.
 * @example
 * "perf:latency:*" will match any namespace starting with "perf:latency:".
 */
```

Note that for `@example` blocks, the line containing the tag is reserved for the title of the example, so we have to put the contents on a separate line to be parsed as we expect.

seanimam replied:
Updated.

Josmithr commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -39,15 +39,29 @@ export interface FluidAppInsightsLoggerConfig {
+	/**
+	 * The namespace pattern to filter telemetry events.
+	 * For example, "perf:latency:*" will match any namespace starting with "perf:latency:"
+	 * This is a regex pattern.
```
The example you gave was in terms of glob-style patterns ("perf:latency:*"), rather than as a RegEx. Is this correct?

seanimam replied:
I'm not entirely sure what the difference is here, if you want to explain but also we moved to `.startsWith()` so this may no longer be a relevant discussion

Josmithr replied:
It's just that the example RegEx didn't handle the case being described. `perf:latency:*` would match the literal string `perf:latency` followed by 0 or more instances of `:`. The desired RegEx would have been `perf:latency:.*` (or `perf:latency:.+` if you wanted to ensure at least one more character after the `:`).

But moving to `startsWith` resolves it :)

Josmithr commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -39,15 +39,29 @@ export interface FluidAppInsightsLoggerConfig {
+	 * events from "perf:latency:ops", you would add "perf.latency.ops" to this list.
+	 * These are regex patterns.
+	 */
+	namespacePatternExceptions?: string[];
```
I'm not sure this needs to be separate from `namespacePattern`. Whether it's a glob pattern or regex, users should be able to specify exceptions inline.

seanimam replied:
Since I swapped from Regex to .startsWith per alejandros suggestions, we'll have to keep this since we'll need an explicit place for exceptions.

Josmithr commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -70,6 +84,15 @@ export class FluidAppInsightsLogger implements ITelemetryBaseLogger {
 	public constructor(client: ApplicationInsights, config?: FluidAppInsightsLoggerConfig) {
 		this.baseLoggingClient = client;
+
+		for (const filter of config?.filtering.filters ?? []) {
```
I'm wondering how necessary this is. If someone specifies an effectively empty filter, won't the effect be that we just no-op?

Josmithr replied:
If it is important that the filters be non-empty, an alternative approach would be to define multiple types, rather than sharing a single interface. Something like:

```typescript
interface CategoryFilter {
  category: TelemetryEventCategory;
}

interface NamespaceFilter {
  namespacePattern: string;
}

type TelemetryFilter = CategoryFilter | NamespaceFilter;
```

This has the added benefit of requiring that all filters you get are "valid" (non-empty).

seanimam replied:
I've updated to this approach, thanks for pointing that out.

Josmithr replied:
No problem! Glad it worked out!

alexvy86 reviewed (commented):
Besides the couple of comments below, I'm still thinking through the approach. 

One thing I noticed: if right now we wanted all log categories for namespace X and only that, we would need to use the mode that lets nothing through by default, and then add 3 separate filters, for the same namespace, each for a different category, right? I think with the introduction of namespaces into the mix, it might make sense to make category an array.

The current approach with namespace and namespace exceptions works as far as I can tell, just wondering how much effort it would be to copy how other libraries/frameworks do it, where the most specific filter (in terms of namespace) that applies to a given event is the one that dictates what happens to it. E.g. if we had a filter for `{ namespace: "A.B.C", categories: ["generic", "error"] }` and one for `{ namespace: "A.B", categories: ["generic"] }`, then any event whose namespace starts with "A.B.C" would match when its category is generic _or_ error; and any event whose namespace only starts with "A.B" would match only when its category is generic. I think this might be doable if we process filters in reverse order of namespace length.

Josmithr commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -104,9 +127,47 @@ export class FluidAppInsightsLogger implements ITelemetryBaseLogger {
+			let matches = true;
+
+			// If the filter has a category and it doesn't match the event's category, skip to next filter
+			if (filter.category && event.category !== filter.category) {
```
Note that this will treat an empty `filter.category` string value the same as if were `undefined`. I suspect this probably isn't what we want. I would expect that if a user specified an empty string as their category, it would match nothing. In which case, this should be:
```suggestion
			if (filter.category !== undefined && event.category !== filter.category) {
```

If that's not the case, and we do want to treat empty string as "no category provided", we should be sure to document that explicitly in the property docs.

seanimam replied:
You are right. After moving to a list of categories this shouldn't be an issue any more.

Josmithr commented on packages/framework/client-logger/app-insights-logger/src/test/fluidAppInsightsLogger.spec.ts:
```diff
@@ -7,19 +7,28 @@ import { strict as assert } from "node:assert";
 import { ApplicationInsights, type IEventTelemetry } from "@microsoft/applicationinsights-web";
```
Loving all of the tests! Thanks!!

Josmithr reviewed (commented):
I left a few pattern-related suggestions that I think would be improvements both for the end-user and for implementation maintenance. Definitely let me know if you have any questions - I'm more than happy to chat in more detail if that would be helpful.

seanimam committed:
Swaps from regex to .startsWith(), updates filter to allow multiple categories, adds more tests, jsdoc updates to use @example tags

seanimam commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -104,9 +127,47 @@ export class FluidAppInsightsLogger implements ITelemetryBaseLogger {
+
+			// If the filter has a namespacePattern, test the event's namespace against it
+			if (filter.namespacePattern !== undefined) {
+				const pattern = new RegExp(filter.namespacePattern);
```
Updated from regex to .startsWith(). I don't know how realistic an attack here would be but it is certainly a possibility and regex is overkill and for this use case.

seanimam committed:
1 letter jsdoc update

seanimam commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -39,15 +39,29 @@ export interface FluidAppInsightsLoggerConfig {
  * Object used with an {@link FluidAppInsightsLoggerConfig}
- * to define logic for filtering of telemetry events
+ * to define a filter with logic for matching it to telemetry events.
+ * Filters can include either a category, namespace or both types of filter; A valid filter must have at least one defined.
```
Updated.

seanimam committed:
Refactors TelemetryFilter interface to ensure either category or namespace filter is provided

seanimam commented:
> Besides the couple of comments below, I'm still thinking through the approach.
> 
> One thing I noticed: if right now we wanted all log categories for namespace X and only that, we would need to use the mode that lets nothing through by default, and then add 3 separate filters, for the same namespace, each for a different category, right? I think with the introduction of namespaces into the mix, it might make sense to make category an array.
> 
> The current approach with namespace and namespace exceptions works as far as I can tell, just wondering how much effort it would be to copy how other libraries/frameworks do it, where the most specific filter (in terms of namespace) that applies to a given event is the one that dictates what happens to it. E.g. if we had a filter for `{ namespace: "A.B.C", categories: ["generic", "error"] }` and one for `{ namespace: "A.B", categories: ["generic"] }`, then any event whose namespace starts with "A.B.C" would match when its category is generic _or_ error; and any event whose namespace only starts with "A.B" would match only when its category is generic. I think this might be doable if we process filters in reverse order of namespace length.

First, I have updated the code to accept an array of categories. Good point.

On the second thought of the most specific filter applying... The current design evaluates all filters until if finds a match.
For the following setup:
```
filters: [
   {
      categories: ["generic", "error"],
      namespace: "A.B.C"
   },
   {
      categories: ["generic"],
      namespace:: "A.B",
   }
]

```

If an event matches the first filter it will be filtered even if it doesn't match the second filter.

It may not be evaluating filters in the most efficient manner but that's something we can adjust later. The most efficient might be some kind of tree like structure given that namespaces have a hierarchy.


> Besides the couple of comments below, I'm still thinking through the approach.
> 
> One thing I noticed: if right now we wanted all log categories for namespace X and only that, we would need to use the mode that lets nothing through by default, and then add 3 separate filters, for the same namespace, each for a different category, right? I think with the introduction of namespaces into the mix, it might make sense to make category an array.
> 
> The current approach with namespace and namespace exceptions works as far as I can tell, just wondering how much effort it would be to copy how other libraries/frameworks do it, where the most specific filter (in terms of namespace) that applies to a given event is the one that dictates what happens to it. E.g. if we had a filter for `{ namespace: "A.B.C", categories: ["generic", "error"] }` and one for `{ namespace: "A.B", categories: ["generic"] }`, then any event whose namespace starts with "A.B.C" would match when its category is generic _or_ error; and any event whose namespace only starts with "A.B" would match only when its category is generic. I think this might be doable if we process filters in reverse order of namespace length.

On the note of categories, I've added the ability to specify an array of categories instead. Good callout.

In regards to your example, the current setup checks against all filters until it finds a match so it shouldn't be a problem. So in that case the configuration:
```
filters: [
     {
       category: ["generic", "error"]
       namespace: "A.B.C"
     },
     {
       category: ["generic"]
       namespace: "A.B"
     },
]
```

1. an event with generic category and namespace A.B.C will match both filters
2. an event with error category and namespace A.B.C will match the first filter

In both cases, the event will be filtered

If in the future matching filters applied something invoked a user defined callback for an event, like a designated transformation or processing step, then this could be an issue (which filter's processing step do you use? -- Most likely you'd want the more specific one as you said). But, for now it should be fine I think.

Josmithr commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -104,9 +137,52 @@ export class FluidAppInsightsLogger implements ITelemetryBaseLogger {
+	/**
+	 * Checks if a given telemetry event conforms to any of the provided {@link TelemetryFilter} rules.
+	 *
+	 * - If a {@link TelemetryFilter} specifies both a `category` and a `namespace`, the event must match both.
```
Note: this won't appear in generated comments the way you expect. It's an unfortunate symptom of how TSDoc handles hard line breaks as soft breaks. In order for these to appear as a list in hover-over / generated API docs, there needs to be a blank line between each list item.

https://github.com/microsoft/FluidFramework/wiki/TSDoc-Guidelines#line-breaks

seanimam replied:
Updated.

alexvy86 commented:
> In regards to your example, the current setup checks against all filters until it finds a match so it shouldn't be a problem. So in your example case the configuration:
> 
> ```
> filters: [
>      {
>        category: ["generic", "error"]
>        namespace: "A.B.C"
>      },
>      {
>        category: ["generic"]
>        namespace: "A.B"
>      },
> ]
> ```
> 
> 1. an event with generic category and namespace A.B.C will match both filters
> 2. an event with error category and namespace A.B.C will match the first filter
> 3. an event with "generic" category and namespace A.B will only match the first filter
> 
> I believe the events here will be filtered as expected.
> 
> If in the future matching filters applied something like invoking a user defined callback for a matching event, like a designated transformation or processing step, then this could be an issue (which filter's processing step do you use? -- Most likely you'd want the more specific one as you said). But, for now it should be fine I think.

True for the example I gave but now consider if the user wants only errors for the more specific namespace, and generic + error in the parent namespace, i.e. `{ namespace: "A.B.C", categories: ["error"] }` and `{ namespace: "A.B", categories: ["generic", "error"] }`. If a generic event in namespace A.B.C gets evaluated by the second filter first, it will be let through even though the more specific filter says it shouldn't have.

alexvy86 commented:
I think I'm convincing myself that it's not too hard to implement that more-specific-filters first approach, which allows for the scenario in my previous comment, on top of what you already have. Just sorting the filters by namespaces length from longest to shortest, once, at startup, and then evaluating them in that order with the same logic you have right now seems like it should be enough. Does it sound reasonable from your end @seanimam ?

seanimam committed:
Refactors filter evaulation and sorting to use the most specific, relevant filter for a given event and adds 2 test cases for the new logic

seanimam commented:
> I think I'm convincing myself that it's not too hard to implement that more-specific-filters first approach, which allows for the scenario in my previous comment, on top of what you already have. Just sorting the filters by namespaces length from longest to shortest, once, at startup, and then evaluating them in that order with the same logic you have right now seems like it should be enough. Does it sound reasonable from your end @seanimam ?

Yes, thanks for clarifying with that example. I've updated the code logic to utilize this methodology and added 2 test cases.

WayneFerrao commented on packages/framework/client-logger/app-insights-logger/src/test/fluidAppInsightsLogger.spec.ts:
```diff
@@ -105,104 +177,546 @@ describe("Telemetry Filter - Category filtering", () => {
+		sinonAssert.callCount(trackEventSpy, 2);
+	});
+
+	// ------------------------------------------------------------------------------------
```
Is this meant to separate a subclass of tests within namespace filter?

alexvy86 commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -30,26 +30,88 @@ export interface FluidAppInsightsLoggerConfig {
+	namespacePatternExceptions?: string[];
 }
 
+/**
```
I'd rewrite the docs for this a bit. On one hand, the logic for picking which filter will apply for a given event; that's the namespace part, and here we also cover that the namespace exceptions allow for more selectivity. The filter with the most specific (longest) namespace pattern that matches an event is the one that gets used.

On the other hand, once the filter for an event is determined, what the filter does is evaluate the category in the event against the ones that are configured in it. And we can add a note that a filter with no namespace is just a catch-all for events that don't match any other filters.

Not for docs, but I imagine a common scenario would be to define specific filters for some namespaces where the user wants to see everything, but for anything not explicitly defined, they only want to see errors, so the last filter would be empty namespace and categories error.

alexvy86 commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -104,12 +174,97 @@ export class FluidAppInsightsLogger implements ITelemetryBaseLogger {
 		return false;
 	}
+
+	public static sortTelemetryFilters(filters: TelemetryFilter[]): TelemetryFilter[] {
```
This can be a non-exported free function in the module, I don't think it need to be exposed or part of the class.

alexvy86 commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -104,12 +174,97 @@ export class FluidAppInsightsLogger implements ITelemetryBaseLogger {
+		});
+	}
+
+	private static getFilterSpecificityScore(filter: TelemetryFilter): number {
```
Following up on the big comment above, I think this only needs to be a check for the namespacePattern length (coalesced to 0). I'd probably inline it when doing the sort.

seanimam replied:
Updated to sort within the constructor using namespace length.

alexvy86 commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -70,12 +132,20 @@ export class FluidAppInsightsLogger implements ITelemetryBaseLogger {
 			},
 		};
+
+		if (this.config.filtering.filters) {
+			// Sorts filters in order from most specific to least specific.
+			this.config.filtering.filters = FluidAppInsightsLogger.sortTelemetryFilters(
+				this.config.filtering.filters,
+			);
+		}
```
Now with the sorting, we probably want to save the sorted copy of the list in a local variable instead of updating the property; the `config` being an object passed by the caller, we probably don't want to mutate it in any way.

In fact... we might want to make a deep copy of it to store locally (and then it's fine to update the filter list in place), so that any changes that the caller might make to the object they pass in don't affect the logger.

seanimam replied:
Agreed. I've updated it to create an internal deep copy filter array.

alexvy86 commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -104,12 +174,97 @@ export class FluidAppInsightsLogger implements ITelemetryBaseLogger {
+					if (doesFilterCategoriesMatch) {
+						// The most specific, relevant filter matches so no need to attempt to evaluate against other filters
+						// as long as the events namespace does not match any defined namespace exception.
+						if (filter.namespacePatternExceptions !== undefined) {
+							for (const patternException of filter.namespacePatternExceptions) {
+								if (event.eventName.startsWith(patternException)) {
+									return false;
+								}
+							}
+						}
```
I'd move this to be the first check once we match on namespace. But note that it cannot `return false`, that prevents any further filters from being evaluated; I think inside the loop we just want to `continue;` so the next filter gets a change to see if it matches the event. That applies to the lines below too.

alexvy86 replied:
Reviewed with Sean, I misread how this works. It's actually correct.

alexvy86 reviewed (commented):
A few comments on the tests, which overall look really good! Great job writing those.

seanimam committed:
Adds filter validation function, changes filter sorting to only use namespace pattern length, filters are deep copied to an internal array, adds getter for logger filter array, merges redundant tests, adds test for fitler validation, pure category filters are now merged internally

seanimam commented on packages/framework/client-logger/app-insights-logger/src/test/fluidAppInsightsLogger.spec.ts:
```diff
@@ -105,104 +177,546 @@ describe("Telemetry Filter - Category filtering", () => {
 
-		const errorCategoryEvent = {
+		// should be excluded - the event matches the second filter but the first filter is more specific to just allow "error"
+		// for the given namespace so it should be evaulated first
```
Updated.

seanimam commented on packages/framework/client-logger/app-insights-logger/src/test/fluidAppInsightsLogger.spec.ts:
```diff
@@ -105,104 +177,546 @@ describe("Telemetry Filter - Category filtering", () => {
+			eventName: "A:B:C",
+		};
+
+		// should be included - matches both filters
```
Updated.

seanimam commented on packages/framework/client-logger/app-insights-logger/src/test/fluidAppInsightsLogger.spec.ts:
```diff
@@ -105,104 +177,546 @@ describe("Telemetry Filter - Category filtering", () => {
+		logger.send(event1);
+		logger.send(event2);
+		sinonAssert.callCount(trackEventSpy, 2);
+	});
```
Merged the tests.

seanimam commented on packages/framework/client-logger/app-insights-logger/src/test/fluidAppInsightsLogger.spec.ts:
```diff
@@ -105,104 +177,546 @@ describe("Telemetry Filter - Category filtering", () => {
+		logger.send(event1);
+		logger.send(event2);
+		sinonAssert.callCount(trackEventSpy, 0);
+	});
```
Merged the tests.

seanimam commented on packages/framework/client-logger/app-insights-logger/src/test/fluidAppInsightsLogger.spec.ts:
```diff
@@ -105,104 +177,546 @@ describe("Telemetry Filter - Category filtering", () => {
+
+	// ------------------------------------------------------------------------------------
+
+	it("inclusive filtering mode DOES SEND events that DO MATCH with atleast one category within a single filter containing multiple categories", () => {
```
For the same reason as my prior comment,  I'd like to keep this logic:
> 1. The test is seeing if matching works as expected against a filter with multiple categories defined. The new logic to merge pure category filters will still result in a single filter with multiple categories so this is a logic branch I'd like to test.

seanimam commented on packages/framework/client-logger/app-insights-logger/src/test/fluidAppInsightsLogger.spec.ts:
```diff
@@ -105,104 +177,546 @@ describe("Telemetry Filter - Category filtering", () => {
 	});
 
-	it("exclusive filter mode sends ALL events except those that match category filters", () => {
+	it("exclusive filtering mode DOES SEND events that DO MATCH with atleast one category within a single filter containing multiple categories", () => {
```
You are right in missing a "NOT". But tests are still relevant.
1. The test is seeing if matching works as expected against a filter with multiple categories defined. The new logic to merge pure category filters will still result in a single filter with multiple categories so this is a logic branch I'd like to test.

seanimam replied:
Updated to throw errors if there is more than one filter that only defines categories.

seanimam commented on packages/framework/client-logger/app-insights-logger/src/test/fluidAppInsightsLogger.spec.ts:
```diff
@@ -90,9 +142,29 @@ describe("Telemetry Filter - filter mode", () => {
+describe("Telemetry Filter - Category Filtering", () => {
 	let appInsightsClient: ApplicationInsights;
 	let trackEventSpy: Sinon.SinonSpy;
+	const configFilters: TelemetryFilter[] = [
+		{
+			categories: ["performance"],
+		},
+		{
+			categories: ["generic"],
+		},
+	];
```
Updated the code to include a validation function for filters which:
1. merges redundant pure category filters (and console.warns)
2. throws error if duplicate namespace filters are provided
3. throws error if namespace pattern exception is provided that is not a child of the parent namespace

seanimam replied:
Per our other conversation, this was updated to throw an error if you define multiple filters  that only define categories rather than merge them internally.

seanimam committed:
Updates namespace pattern exceptions to a set instead of an array

alexvy86 commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -76,6 +143,22 @@ export class FluidAppInsightsLogger implements ITelemetryBaseLogger {
 		};
+
+		// Deep copy filters to internal array
+		this._filters =
+			this.config.filtering.filters?.map((filter) => structuredClone(filter)) ?? [];
```
Let's copy the whole structure, not just the filters. Right now if the user changes the mode on the config object they passed in, that would have a runtime effect because `shouldSendEvent()` cehcks `this.config.filtering.mode`.

seanimam replied:
Updated to deep copy the entire config.

seanimam committed:
Updates filter validation to only allow a single pure category filter and updates jsdoc and related tests. Updates jsdoc to more clearly explain that providing a filter with no categories the defined is the same as defining all categories.

alexvy86 commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -133,32 +134,27 @@ export class FluidAppInsightsLogger implements ITelemetryBaseLogger {
-			return namespaceBLength - namespaceALength;
-		});
-	}
+		// Deep copy config to internal array
```
```suggestion
		// Deep copy config to prevent issues if user mutates the object they passed in
```

seanimam replied:
Updated

alexvy86 commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -133,32 +134,27 @@ export class FluidAppInsightsLogger implements ITelemetryBaseLogger {
+			this.validateFilters(this.config.filtering.filters);
+			// Sort filters by longest namespace first.
+			this.config.filtering.filters.sort((a, b) => {
+				const namespaceALength = "namespacePattern" in a ? a.namespacePattern.length : 0;
+				const namespaceBLength = "namespacePattern" in b ? b.namespacePattern.length : 0;
```
It achieves the same thing but I think it's a more common pattern in our codebase. It also has the advantage that refactoring tools will detect it. For the same reason, let's update `validateFilters()` to also do undefined checks instead of property existence checks with strings.
```suggestion
				const namespaceALength = a.namespacePattern !== undefined ? a.namespacePattern.length : 0;
				const namespaceBLength = b.namespacePattern !== undefined ? b.namespacePattern.length : 0;
```

seanimam replied:
`a.namespacePattern !== undefined` seems to no type cast the objects but using the `"namespacePattern" in a` does. Let me know if I'm missing something there.

This is the specific error:
```
Property 'namespacePattern' does not exist on type 'TelemetryFilter'.
  Property 'namespacePattern' does not exist on type 'CategoryFilter'
```

alexvy86 replied:
Ah, we have the type union... ok, yeah, for now keep it and let's create an ADO task to review this at some point (not necessarily change it, but we should have a think).

seanimam replied:
Per our conversation I have updated to do a combination of both, `"namespacePattern" in a && a.namespacePattern !== undefined`

alexvy86 commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -276,40 +271,20 @@ export class FluidAppInsightsLogger implements ITelemetryBaseLogger {
-						console.warn("Redundant category filters found. Filters will be merged.");
-					}
-					pureCategoryFilters.add(category);
+				if (hasPureCategoryFilter) {
+					throw new Error("Cannot have multiple filters that only define categories");
```
We can validate these in the same way that the others by adding an entry with empty namespace to `uniqueFilterNamespaces`, which also matches how we should think of them (filters for the default namespace).

seanimam replied:
Updated to this.

alexvy86 commented on packages/framework/client-logger/app-insights-logger/src/test/fluidAppInsightsLogger.spec.ts:
```diff
@@ -70,89 +70,29 @@ describe("FluidAppInsightsLogger", () => {
-		debugger;
-
-		assert.deepStrictEqual(logger.filters, expectedSortedFilterArray);
+		assert.throws(() => new FluidAppInsightsLogger(appInsightsClient, invalidConfig));
```
Nit (let's create a follow up task for it): when asserting that something throws, it's a good practice to check that the thrown error is what we expect. Otherwise we might introduce a change in behavior that causes a different problem, and we wouldn't know it with this test.

seanimam replied:
I've gone ahead and updated the tests to assert for a specific error and message combination. In the future we may want to consider standardizing the error messages

seanimam committed:
Updates typecasting to explicit check for undefined, small logic updates to filter validation function, tests involving throwing an error now check for explicit error messages

seanimam committed:
adds more explicit undefined checks for filter.categories and filter.namespacePattern

Josmithr commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -30,26 +30,93 @@ export interface FluidAppInsightsLoggerConfig {
 		/**
-		 * Controls the default filtering of log events by their category.
+		 * Controls the filtering of log events.
 		 * Leaving this undefined will be treated as an empty array.
```
Nit: I would put this part of the comment in a `@remarks` block, since it describes usage notes, rather than a semantic description.

seanimam replied:
Updated.

Josmithr commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -30,26 +30,93 @@ export interface FluidAppInsightsLoggerConfig {
 		 * Leaving this undefined will be treated as an empty array.
+		 *
+		 * In order for the filters to be valid they must meet the following conditions:
+		 * 1. There must not be any two filters with the same `namespacePattern`.
```
Remember that you'll need blank lines between entries here in order to have this treated as a list by TSDoc. You'll also need a blank line before `1.`, or the first list entry will appear on the same line as the preceding text.

Josmithr replied:
```suggestion
		 *
		 * 1. There must not be any two filters with the same `namespacePattern`.
```

seanimam replied:
Updated.

Josmithr commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -30,26 +30,93 @@ export interface FluidAppInsightsLoggerConfig {
-	 * The category of telemetry event that this filter applies to
+	 * A list of namespace patterns to explicitly exclude from the filter.
+	 *
+	 * @example If you have a namespacePattern of "perf:latency" but want to exclude
```
```suggestion
	 * @example
	 If you have a namespacePattern of "perf:latency" but want to exclude
```

Otherwise the text on this line will be treated as the example's title, rather than contents of it.

seanimam replied:
Updated.

Josmithr commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -30,26 +30,93 @@ export interface FluidAppInsightsLoggerConfig {
+ *
+ * @public
+ *
+ * @example With the following configuration, an event `{ namespace: "A.B.C", categories: ["generic"] }` will not be sent despite matching the first, less specific filter because it did not match the second filter which was the most relevant and specific
```
```suggestion
 * @example
 * With the following configuration, an event `{ namespace: "A.B.C", categories: ["generic"] }` will not be sent despite matching the first, less specific filter because it did not match the second filter which was the most relevant and specific
```

seanimam replied:
Updated.

Josmithr commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -104,12 +189,118 @@ export class FluidAppInsightsLogger implements ITelemetryBaseLogger {
+	 *
+	 * @param event - The telemetry event to check against the filters.
+	 *
+	 * @returns boolean `true` if the event matches any filter, otherwise `false`.
```
```suggestion
	 * @returns `true` if the event matches any filter, otherwise `false`.
```

seanimam replied:
Updated.

Josmithr commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -104,12 +189,118 @@ export class FluidAppInsightsLogger implements ITelemetryBaseLogger {
+	/**
+	 * Checks an array of telemetry filters for any issues, merges redundant filters, and returns a fully validated array.
+	 *
+	 * @throws - an Error if there are two filters with duplicate namespace patterns or a filter with a pattern exception that is not a child of the parent pattern.
```
```suggestion
	 * @throws An error if there are two filters with duplicate namespace patterns or a filter with a pattern exception that is not a child of the parent pattern.
```

Josmithr replied:
Generally, only the `@param` comments want a `-` in their syntax. Including a hyphen here will result in it appearing as a string literal in the docs.

seanimam replied:
Updated.

Josmithr commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -30,26 +30,93 @@ export interface FluidAppInsightsLoggerConfig {
+	 * A list of namespace patterns to explicitly exclude from the filter.
+	 *
+	 * @example If you have a namespacePattern of "perf:latency" but want to exclude
+	 *
```
```suggestion
```

seanimam replied:
Updated

Josmithr commented on packages/framework/client-logger/app-insights-logger/src/fluidAppInsightsLogger.ts:
```diff
@@ -30,26 +30,93 @@ export interface FluidAppInsightsLoggerConfig {
+		 *
+		 * In order for the filters to be valid they must meet the following conditions:
+		 * 1. There must not be any two filters with the same `namespacePattern`.
+		 * 2. All {@link NamespaceFilter} must not have any defined `namespacePatternException` that is not a child of the parent `namespacePattern`
```
```suggestion
		 *
		 * 2. All {@link NamespaceFilter} must not have any defined `namespacePatternException` that is not a child of the parent `namespacePattern`
```

seanimam replied:
Updated

Josmithr reviewed (approved):
Left a few more TSDoc syntax comments. Otherwise, looks good!

seanimam committed:
jsdoc tweaks
